### PR TITLE
pkgs/ok_http: JNIgen fixes and added WebSocket support

### PR DESCRIPTION
**Pull request by Anikate-De**
Items in this PR:
- [x] Update `package: jni, jnigen` versions 
- [x] Several `jnigen` usage fixes (thanks to @HosseinYousefi)
- [x] Generate JNI bindings for `WebSocket` support required classes 
- [x] Implement `WebSocket`
- [x] Pass all conformance tests in `package: web_socket_conformance_tests`
- [x] Modify the example app to showcase both HTTP and WebSocket side-by-side
- [x] Added Docs
- [x] Updated README.md to boast new WebSocket support
- [x] Updated CHANGELOG.md

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/blob/main/docs/External-Package-Maintenance.md#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by Anikate-De** [comment]
Requesting for review by pinging as I don't have permissions
@brianquinlan @camsim99 @HosseinYousefi

File: pkgs/ok_http/lib/src/ok_http_client.dart
Comment (HosseinYousefi) [inline pkgs/ok_http/lib/src/ok_http_client.dart:291]: This can be done without a for loop. `JArray<jbyte>` also has a `getRange` method that you can use and it will be faster.

File: pkgs/ok_http/README.md
Comment (camsim99) [inline pkgs/ok_http/README.md:28]: Could you include what it will allow you to do in your Flutter app with ease?

File: pkgs/ok_http/example/lib/main.dart
Comment (camsim99) [inline pkgs/ok_http/example/lib/main.dart:201]: Nit: can this be a map or some data class you make for clarity?

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (camsim99) [inline pkgs/ok_http/lib/src/ok_http_web_socket.dart:114]:
```suggestion
              _events.add(CloseReceived(1006, 'Closed abnormally'));
```

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (camsim99) [inline pkgs/ok_http/lib/src/ok_http_web_socket.dart:180]: Nit (mostly because I looked this up): maybe link to https://square.github.io/okhttp/5.x/okhttp/okhttp3/-ok-http-client/index.html or where this is mentioned?

**Review by camsim99 — APPROVED** [review]
Looks good to me overall! Was able to understand this implementation by following the WebSocket docs :) just left some nits

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (brianquinlan) [inline pkgs/ok_http/lib/src/ok_http_web_socket.dart:45]: Is there a reason why you created an empty constructor and do the work in the `_connect` method instead of having the same logic in `connect`?

I'm asking because it might be since if the approach is the same as is used by [`BrowserWebSocket`](https://github.com/dart-lang/http/blob/4178b67e7970ff02926e79649db5ab473b46cf30/pkgs/web_socket/lib/src/browser_web_socket.dart#L28)

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (brianquinlan) [inline pkgs/ok_http/lib/src/ok_http_web_socket.dart:71]: `releaseOriginal: true` ?

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (brianquinlan) [inline pkgs/ok_http/lib/src/ok_http_web_socket.dart:75]: Won't this get formatted weirdly if `protocols` is `null`?

**Review by brianquinlan — CHANGES_REQUESTED** [review]
Amazing work!

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (HosseinYousefi) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:71]: @brianquinlan @Anikate-De would it be nicer for you to use auto releasing zones? Please give me your feedback here: https://github.com/dart-lang/native/issues/702#issuecomment-2154670370

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:79]: I guess not? If `protocols` is `null`, then the condition in the `if` block is `false` and hence the statements inside will not be executed.

Only when `protocols` is non-null and it does not contain the received `protocol` shall we throw the error.

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart]: Done! Following https://github.com/dart-lang/native/issues/702#issuecomment-2218813253, I think it wouldn't be much better to use auto-releasing zones.

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:49]: Yes! Since we need an instance of `bindings.OkHttpClient` to both open and close WebSockets, we cannot access it statically inside `connect()`, which I had to use to make the implementation more conformant with the other WebSockets.

Inside `static connect()`, we make a new instance of the OkHttpClient and then create a non-static `_connect` method that can access this instance. 

(I added some docs to explain the gist of it)

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:184]: Done!

File: pkgs/ok_http/example/lib/main.dart
Comment (Anikate-De) [reply pkgs/ok_http/example/lib/main.dart]: Done, removed the WebSocket example overall. I will add it to the main `README`, and docs, as in https://github.com/dart-lang/http/pull/1266.

File: pkgs/ok_http/android/src/main/kotlin/com/example/ok_http/WSInterceptor.kt
Comment (Anikate-De) [reply pkgs/ok_http/android/src/main/kotlin/com/example/ok_http/WSInterceptor.kt]: Linked to https://github.com/dart-lang/native/issues/1337.

File: pkgs/ok_http/example/lib/main.dart
Comment (Anikate-De) [reply pkgs/ok_http/example/lib/main.dart]: Not proceeding, as we are removing the WebSocket example overall.

File: pkgs/ok_http/lib/src/ok_http_client.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_client.dart]: Changed it to private, but needed to add one in `ok_http_web_socket.dart` as well.

File: pkgs/ok_http/lib/src/ok_http_client.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_client.dart]: Changed it to private, but needed to add one in `ok_http_web_socket.dart` as well.

File: pkgs/ok_http/android/src/main/kotlin/com/example/ok_http/WSInterceptor.kt
Comment (Anikate-De) [reply pkgs/ok_http/android/src/main/kotlin/com/example/ok_http/WSInterceptor.kt]: Done!

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:184]: Not proceeding as the WebSocket example has been removed.

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart]: Done! Used `abnormal close`

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:184]: Done!

File: pkgs/ok_http/example/lib/main.dart
Comment (Anikate-De) [reply pkgs/ok_http/example/lib/main.dart]: Not proceeding as the WebSocket example has been removed.

File: pkgs/ok_http/lib/src/ok_http_client.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_client.dart]: Done! Thanks!

File: pkgs/ok_http/README.md
Comment (Anikate-De) [reply pkgs/ok_http/README.md]: Done! Added a list of bullet points that showcases OkHttp features, and makes it easy to add more to the list. Left the table as it is (it looks cool, flexing over all other clients :) )

File: pkgs/ok_http/README.md
Comment (Anikate-De) [reply pkgs/ok_http/README.md]: Done! Added exclusion rule to my markdown formatter, won't force table formats again!

**Review by camsim99 — APPROVED** [review]
LGTM! Just left some nits

File: pkgs/ok_http/README.md
Comment (Anikate-De) [reply pkgs/ok_http/README.md:34]: That's a nice point! Here the example is as minimal/boring as it can get, mostly to demonstrate how conformant its usage is, in comparison to other WebSockets. See PR https://github.com/dart-lang/http/pull/1266 that added the same example to `CupertinoWebSocket`, which was originally taken from the `README` of `package: web_socket`. This package contains a better example of WebSockets (Cryptocurrency API)

If you still consider that a more comprehensive example is suitable for this as well, I can start working on it!

File: pkgs/ok_http/README.md
Comment (Anikate-De) [reply pkgs/ok_http/README.md:33]: Done!

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart:18]: Done!

**Comment by Anikate-De** [comment]
@camsim99, should I change the version from `ok_http: 0.1.0-wip` to `0.1.0`?

**Comment by camsim99** [comment]
> @camsim99, should I change the version from `ok_http: 0.1.0-wip` to `0.1.0`?

Ah yes because this finishes the implementation.

**Comment by camsim99** [comment]
> Done! Ready for release 🎉

Woohoo!! Looks like something with the CLA happened and I think @HosseinYousefi has to approve, but in the meantime I will follow up on https://github.com/dart-lang/http/issues/1259

**Comment by HosseinYousefi** [comment]
I accepted the CLA with my other email. In general, no need to add me as a co-author when I review and if you do make sure it's my google email. Thanks!

File: pkgs/ok_http/pubspec.yaml
Comment (Anikate-De) [reply pkgs/ok_http/pubspec.yaml]: Done!

File: pkgs/ok_http/lib/src/ok_http_web_socket.dart
Comment (Anikate-De) [reply pkgs/ok_http/lib/src/ok_http_web_socket.dart]: Done!

**Review by HosseinYousefi — APPROVED** [review]
LGTM! Thanks @Anikate-De!

**Comment by HosseinYousefi** [comment]
I dismissed @brianquinlan's review as he's OOO. @camsim99 Feel free to merge this whenever you see fit.

**Comment by camsim99** [comment]
It appears I do not have write access :/ Will need to get this resolved to merge this.

**Pull request merged**
